tools/c7n_azure/c7n_azure/resources/access_control.py

```diff
@@ -96,3 +150,78 @@ class UserRole(RelatedResourceFilter):
 
     RelatedResource = "c7n_azure.resources.access_control.RoleDefinition"
     RelatedIdsExpression = "properties.roleDefinitionId"
+
+
+@RoleAssignment.filter_registry.register('resource-access')
+class ResourceAccessFilter(RelatedResourceFilter):
+    """Filters role assignments that have access to a certain
+    type of azure resource.
+
+    :Example:
+
+        .. code-block:: yaml
+
+            policies:
+               - name: assignments-by-azure-resource
+                 resource: azure.roleassignment
+                 filters:
+                    - type: resource-access
+                      relatedResource: azure.vm
+    """
+
+    schema = type_schema(
+        'resource-access',
+        relatedResource={'type': 'string'},
+        rinherit=RelatedResourceFilter.schema,
+        required=['relatedResource']
+    )
+
+    def __init__(self, data, manager=None):
+        super(ResourceAccessFilter, self).__init__(data, manager)
+        resource_type = self.data['relatedResource']
+        self.factory = Azure.resources.get(
+            resource_type.rsplit('.', 1)[-1])
+
+    def get_related(self, resources):
```

Comment:
fwiw if you just want `get_related` to work correctly its pretty much just an update to the last line to do a contains check instead of an equals

```

    def get_related(self, resources):
        resource_manager = self.get_resource_manager()
        related_ids = self.get_related_ids(resources)
        related = resource_manager.resources()
        return {r['id']: r for r in related
                if any(scope in r['id'] for scope in related_ids)}
```

And set `RelatedIdsExpression = "properties.scope"` in the root of your class to the base `get_related_ids` works right.

Reply:
There isn't a scope on the related resources though. So the line doesn't work. I'm also not sure why you think my implementation doesn't work correctly? 
```        return {r['id']: r for r in related
                if any(scope in r['id'] for scope in related_ids)}
```

Reply:
the scope is enumerator over the `related_ids` so that should check if the resource (VM) ID contains *any* of the related_id's (scopes).  It does the same check you did in process resources basically just keeps the functions doing what they used to do instead of repurposing return types.

Was worried you dropped a lot of code from process_resource, was it important code? hah - I admit I can't quite tell.  It seems like the and/or stuff is relevant.

Reply:
Yeah so that's what I discovered on Tuesday is those other scenarios don't even work in this scenario so I can cut them out. the match-resource stuff isn't relevant and could never happen on this related filter because what it does is dynamically change the match value to the entire resource. Since the main resource is `RoleAssignments` there will never be a case that you can match another resource perfectly to `RoleAssignments` since the validation doesn't even let you do a related filter on them.

Reply:
The `AnnotationKey` stuff is just used for logging it looks like, and I can add that back if we want to log the matches.

Reply:
I couldn't figure out what on Earth that did (match resource).  Think I get it now.